### DISCUSSION TITLE
Using File2ByteSlice for WAV files

### DISCUSSION THREAD

**Discussion by bjazmoore** [hajimehoshi/ebiten#1515] [Q&A] [answered]
I found the article regarding some of the tools that were developed along with Ebiten.  Very interesting in File2byteslice.  I downloaded the source and was able to build an executable.  It works great!  I am encoding image files (PNG's) to byte slices.  I understand how to convert these to an image for using in Ebiten from reading some of the example code.  I am curious - if I encode a WAV file to a byte slice will I be able to use it in the game engine?  How?

Thanks in advance.  P.S. - if I should be taking my basic questions somewhere else - please let me know.

**Comment by hajimehoshi** [comment, marked as answer]
Now I recommend to use Go 1.16's new feature: [embed](https://golang.org/pkg/embed/) instead of file2byteslice. This is the standard way to embed resources and much easier to use. Ebiten will keep using file2byteslice for a while for old Go compilers, but will migrate this into embed in near future.

> if I encode a WAV file to a byte slice will I be able to use it in the game engine? How?

Wav file format can be decoded with [Ebiten's wav package](https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2/audio/wav). See [examples/wav](https://github.com/hajimehoshi/ebiten/blob/master/examples/wav/main.go).

> Thanks in advance. P.S. - if I should be taking my basic questions somewhere else - please let me know.

I welcome any questions related to Ebiten here :-)

**Reply by AWachtendorf** [reply]
I tried encoding a 3 - minute WAV file once. As WAV is one of the complexer audio formats, my byte slice was a friggin 80mb large! The IDE (working with intellij Goland) prompty told me I was crazy for usign this large code files. The only way I could do it was by using an low bit mp3 and making the soundfile 2 minutes shorter. Still this didn't achieve anything regarding usability.  
So to answer the question by my experience: 
Can you convert sound to a byte slice? YES.
Should you? Most likely NO. 
:)

**Reply by hajimehoshi** [reply]
> As WAV is one of the complexer audio formats

The reason why a wav file tends to be big is not because the format is complex, but because the format doesn't compress the data :-) The format itself is pretty simple.

I agree with the conclusion: we should use MP3 or Ogg/Vorbis instead of Wav unless the audio is short enough.

**Reply by AWachtendorf** [reply]
> The reason why a wav file tends to be big is not because the format is complex, but because the format doesn't compress the data :-) The format itself is pretty simple.

Oh, absolutely right! Thank you for correcting me :) I guess that's the reason why sound engineers prefer wav, as its a good format to work with and compress and scale down afterwards, am I correct with that?

**Reply by hajimehoshi** [reply]
I'm not a sound engineer, but I think that's correct, as compressing data loses the original data and results in less quality sound.

**Reply by bjazmoore** [reply]
Thank you for the excellent reply.  I will look at embed.  I also appreciate the pointer.  Also appreciate the discussion on compressed file formats for audio.  

These comments answer my question.  Thank you.
